include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -1578,6 +1585,1157 @@ namespace Differentiation
+         */
+        static const unsigned int rank = SymmetricTensor<2, dim>::rank;
+
+        /**
+         * The tensor type associated with this field.
+         */
+        template <typename NumberType>
+        using tensor_type = SymmetricTensor<rank, dim, NumberType>;
+
+        /**
+         * The value type associated with this field.
+         */
+        template <typename NumberType>
+        using value_type = tensor_type<NumberType>;
+
+        /**
+         * The gradient type associated with this field.
+         */
+        template <typename NumberType>
+        using gradient_type = Tensor<rank + 1, dim, NumberType>;
+
+        /**
+         * Return the table indices corresponding to
+         * @p column_offset entry of the input @p table_indices.
+         */
+        template <int rank_in>
+        static TableIndices<rank>
+        table_index_view(const TableIndices<rank_in> &table_indices,
+                         const unsigned int           column_offset)
+        {
+          Assert(column_offset < rank_in, ExcInternalError());
+          Assert(column_offset + 1 < rank_in, ExcInternalError());
+          return TableIndices<rank>(table_indices[column_offset],
+                                    table_indices[column_offset + 1]);
+        }
+
+        /**
+         * Return the first global component of this field.
+         */
+        static inline unsigned int
+        first_component(const FEValuesExtractors::SymmetricTensor<2> &extractor)
+        {
+          return extractor.first_tensor_component;
+        }
+
+        /**
+         * Return the local unrolled component corresponding to
+         * @p column_offset entry of the @p table_indices.
+         */
+        template <typename IndexType = unsigned int, int rank_in>
+        static IndexType
+        local_component(const TableIndices<rank_in> &table_indices,
+                        const unsigned int           column_offset)
+        {
+          using TensorType = tensor_type<double>;
+          return TensorType::component_to_unrolled_index(
+            table_index_view(table_indices, column_offset));
+        }
+
+        /**
+         * Return a flag that indicates if the input index
+         * corresponds to a symmetric component of the field.
+         */
+        static bool
+        symmetric_component(const unsigned int index)
+        {
+          static const SymmetricTensor<2, dim> t;
+          const TableIndices<2>                table_indices =
+            t.unrolled_to_component_indices(index);
+          return table_indices[0] != table_indices[1];
+        }
+      };
+
+
+      /**
+       * A helper struct that defines the return type of gradient (first
+       * derivative) calculations of scalar fields with respect to a field
+       * defined by the @p ExtractorType template parameter.
+       */
+      template <int dim, typename NumberType, typename ExtractorType>
+      struct Gradient
```

Comment:
> And would using the term Jacobian be less ambiguous here?

I would say that "Jacobian" as a substitute for "Hessian" is more ambiguous, as the construction of [the Jacobian matrix](https://en.wikipedia.org/wiki/Jacobian_matrix_and_determinant) only involves first derivatives. I think that the way that I define it, `Gradient` being the first derivative of a scalar function and `Hessian` being the second derivative is correct and most easy to follow. What I will do though is rename them to `ScalarFieldGradient` and `ScalarFieldHessian`, and then typedef two counterparts namely `VectorFieldValue` and `VectorFieldJacobian` based off of them (since they share the same `type`). This should make their use cases quite unambigious.

Reply:
I was thinking of using "Jacobian" instead of "Gradient". If you name the second derivatives after a guy, maybe also the first?

Though I recognize that we also use your convention in `FEValues`, so maybe just using what you are already doing is the right thing.

Reply:
I best refer to the description of [these functions](https://github.com/dealii/dealii/blob/master/include/deal.II/differentiation/ad/ad_drivers.h#L414-L499) :-D I suppose at this point I'd prefer to consistently use the chosen terminology/convention, and the (revised) version of the naming scheme for these classes is now consistent with our previous definitions.